apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DatabaseManager.java

```diff
@@ -150,38 +160,105 @@ public Jdbi getJdbi() {
     }
 
     /**
-     * Method gives ability to create new 'shard database', open existing shard and add it into shard list.
-     * @param shardId shard name to be added
-     * @return shard database connection pool instance
+     * {@inheritDoc}
      */
     @Override
     public TransactionalDataSource createAndAddShard(Long shardId) {
-        Objects.requireNonNull(shardId, "shardId is NULL");
-        String shardName = ShardNameHelper.getShardNameByShardId(shardId); // convert shard Id into shard name
-        log.debug("Create new SHARD '{}'", shardName);
+        ShardDataSourceCreateHelper shardDataSourceCreateHelper =
+                new ShardDataSourceCreateHelper(this, shardId).createUninitializedDataSource();
+        TransactionalDataSource shardDb = shardDataSourceCreateHelper.getShardDb();
+        shardDb.init(new AplDbVersion());
+        connectedShardDataSourceMap.put(shardDataSourceCreateHelper.getShardId(), shardDb);
+        log.debug("new SHARD '{}' is CREATED", shardDataSourceCreateHelper.getShardName());
+        return shardDb;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public TransactionalDataSource createAndAddShard(Long shardId, DbVersion dbVersion) {
+        Objects.requireNonNull(dbVersion, "dbVersion is null");
+        ShardDataSourceCreateHelper shardDataSourceCreateHelper =
+                new ShardDataSourceCreateHelper(this, shardId).createUninitializedDataSource();
+        TransactionalDataSource shardDb = shardDataSourceCreateHelper.getShardDb();
+        shardId = shardDataSourceCreateHelper.getShardId();
+        if (connectedShardDataSourceMap.containsKey(shardId)) {
+            TransactionalDataSource dataSource = connectedShardDataSourceMap.get(shardId);
+            if (dataSource != null && !dataSource.isShutdown()) {
```

Comment:
No, we don't need throws, because we create Shard db/file with 'tables only schema' for the first time. Later when block/tr data is copied into it from main db we close shard db,  improve it by updating 'constraints + indexes' schema, reopen and replace it in cache by fully initialized version.